internal/cli/cli.go

```diff
@@ -0,0 +1,171 @@
+		s3BackendBucket                   = app.Flag("s3-backend-bucket", "Bucket name to use for the S3 state backend").String()
+		s3BackendPath                     = app.Flag("s3-backend-path", "Path to use when storing state in the S3 state backend").String()
+		debug                             = app.Flag("debug", "Enable debug logging").Bool()
+	)
+
+	app.Action(func(_ *kingpin.ParseContext) error {
+		if newLogger == nil {
+			newLogger = defaultLogger
+		}
+		if newAWSClient == nil {
+			newAWSClient = defaultAWSClientFactory
+		}
+		logger, err := newLogger(*debug)
+		if err != nil {
+			panic(fmt.Errorf("initialize zap logger: %s", err))
+		}
+		defer logger.Sync()
+
+		providers := []sidecred.Provider{random.New(time.Now().UnixNano())}
+
+		if *stsProviderEnabled {
+			_, client, _, _ := newAWSClient()
+			providers = append(providers, sts.New(client,
+				sts.WithExternalID(*stsProviderExternalID),
+				sts.WithSessionDuration(*stsProviderSessionDuration),
+			))
+		}
+
+		if *githubProviderEnabled {
+			privateKey, err := ioutil.ReadFile(*githubProviderPrivateKey)
+			if err != nil {
+				logger.Fatal("read private key", zap.Error(err))
+			}
+			app, err := github.NewAppsClient(*githubProviderIntegrationID, string(privateKey))
+			if err != nil {
+				logger.Fatal("initialize github app", zap.Error(err))
+			}
+			providers = append(providers, github.New(app,
+				github.WithDeployKeyRotationInterval(*githubProviderKeyRotationInterval),
+			))
+		}
+
+		var store sidecred.SecretStore
+		switch sidecred.StoreType(*secretStoreBackend) {
+		case sidecred.SecretsManager:
+			_, _, _, client := newAWSClient()
+			store = secretsmanager.New(client,
+				secretsmanager.WithPathTemplate(*secretsManagerStorePathTemplate),
+			)
+		case sidecred.SSM:
+			_, _, client, _ := newAWSClient()
+			store = ssm.New(client,
+				ssm.WithPathTemplate(*ssmStorePathTemplate),
+				ssm.WithKMSKeyID(*ssmStoreKMSKeyID),
+			)
+		case sidecred.Inprocess:
+			store = inprocess.New(
+				inprocess.WithPathTemplate(*inprocessStorePathTemplate),
+			)
+		default:
+			logger.Fatal("unknown secretstore backend", zap.String("backend", *secretStoreBackend))
+		}
+
+		var backend sidecred.StateBackend
+		switch *stateBackend {
+		case "file":
+			backend = file.New(*fileBackendPath)
+		case "s3":
+			client, _, _, _ := newAWSClient()
+			backend = s3.New(client, *s3BackendBucket, *s3BackendPath)
+		default:
+			logger.Fatal("unknown state backend", zap.String("backend", *stateBackend))
+		}
+
+		s, err := sidecred.New(providers, store, backend, logger)
+		if err != nil {
+			logger.Fatal("initialize sidecred", zap.Error(err))
+		}
+		return run(s)
+	})
+}
+
+func defaultAWSClientFactory() (s3.S3API, sts.STSAPI, ssm.SSMAPI, secretsmanager.SecretsManagerAPI) {
+	var (
+		sess *session.Session
+		err  error
+		once sync.Once
+	)
+	once.Do(func() {
+		sess, err = session.NewSession(&aws.Config{Region: aws.String(os.Getenv("AWS_REGION"))})
```

Comment:
Have you looked into `session.Must` see [reference](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#Must)

Reply:
I tend to not use `session.Must` because I prefer to add a bit of context to the error before it is handled (via panic or otherwise). In this particular case I think it is useful to have the context there because the factory is passed to another function and can be called at an arbitrary time (or not at all), which I think makes the context useful for traceability. But great suggestion non the less!